swifttool/ring_defintion.py

```diff
@@ -0,0 +1,264 @@
+        if data:
+            self.__dict__.update(data)
+
+    def __del__(self):
+        shutil.rmtree(self.workspace)
+
+    def __repr__(self):
+        return str(self.__dict__)
+
+    def _ring_create_command(self, ringtype):
+        return "%s %s/%s.builder create %d %d %d" % (
+            self.ring_builder_cmd, self.workspace, ringtype,
+            int(self.part_power), int(self.replicas),
+            int(self.min_part_hours))
+
+    def _ring_add_command(self, ringtype, zone, host, port, disk,
+                          metadata, weight):
+        return "%s %s/%s.builder add %s-%s:%d/%s_%s %d" % (
+            self.ring_builder_cmd, self.workspace, ringtype, zone, host,
+            int(port), disk, metadata, int(weight))
+
+    def _ring_rebalance_command(self, ringtype):
+        return "%s %s/%s.builder rebalance" % (
+            self.ring_builder_cmd, self.workspace, ringtype)
+
+    def _ring_setweight_command(self, ringtype, zone, host, port,
+                                disk, weight):
+        return "%s %s/%s.builder set_weight %s-%s:%d/%s %d" % (
+            self.ring_builder_cmd, self.workspace, ringtype, zone, host,
+            int(port), disk, int(weight))
+
+    def _ring_remove_command(self, ringtype, zone, host, port, disk):
+        return "%s %s/%s.builder remove %s-%s:%d/%s" % (
+            self.ring_builder_cmd, self.workspace, ringtype, zone, host,
+            int(port), disk)
+
+    def _ring_search_command(self, ringtype, zone, host, port, disk):
+        return "%s %s/%s.builder search %s-%s:%d/%s" % (
+            self.ring_builder_cmd, self.workspace, ringtype, zone, host,
+            int(port), disk)
+
+    @property
+    def nodes(self):
+        ret = set()
+        if self.zones and isinstance(self.zones, dict):
+            for zone, nodes in self.zones.iteritems():
+                ret.update(nodes.keys())
+        return ret
+
+    def generate_commands(self, rebalance=True, meta=None):
+        commands = []
+
+        for ringtype in RING_TYPES:
+            builder_present = os.path.exists("%s/%s.builder" %
+                                             (self.workspace, ringtype))
+            if not builder_present:
+                commands.append(self._ring_create_command(ringtype))
+
+            for zone, nodes in self.zones.iteritems():
+                for node, disks in nodes.iteritems():
+
+                    ringdisks = []
+                    # Add all disks designated for ringtype
+                    if isinstance(disks['disks'], dict):
+                        if ringtype in disks['disks']:
+                            ringdisks += disks['disks'][ringtype]
+                        else:
+                            raise Exception("Malformed ring_defintion.yml. "
+                                            "Unknown ringtype: %s" % ringtype)
+                    elif isinstance(disks['disks'], list):
+                        ringdisks = disks['disks']
```

Comment:
Do we have / need an else block to handle if `disks['disks']` is neither a dict or a list?